### ISSUE TITLE
bug(sdk): tls strategy none

### ISSUE THREAD

**Issue by grutt** [hatchet-dev/hatchet#257]
go client needs support  HATCHET_CLIENT_TLS_STRATEGY=none
```
panic: could not load TLS config: invalid TLS strategy: none 
```

**Comment by steebchen** [comment]
How does this relate to these though?

https://github.com/hatchet-dev/hatchet/blob/78685d0098123178f600b646ae21bd47bc138802/internal/config/loader/loaderutils/tls.go#L22-L32

https://github.com/hatchet-dev/hatchet/blob/78685d0098123178f600b646ae21bd47bc138802/internal/config/loader/loaderutils/tls.go#L44-L60

**Comment by Flo4604** [comment]
> How does this relate to these though?

I don't fully understand what you mean by this, should't TLS none also be available for the go client since its both supported for the typescript and python sdk?

When hosting hatchet locally and wanting to connect with go and without tls, having the none option support would be nice.

**Comment by steebchen** [comment]
You are right @Flo4604; I'm just not sure why there is no handling of the `none` value in the engine at this time. @grutt wouldn't we need to set it up here first before adding it in the SDKs or what am I missing?

**Comment by Flo4604** [comment]
please correct me If im wrong but the none strategy doesn't have to be implemented on the server when	 `SERVER_GRPC_INSECURE: "t"` is set. 

In the docker compose file the `SERVER_TLS_STRATEGY` is never overwritten so it defaults to tls

Since then tls is not used, the go client currently has issues with that as It tries to use tls / mtls for a server which doesn't use either.

**Comment by mbecker** [comment]
Hi,

follwing the official Docker Compose instructions / tutorial at the following page [https://docs.hatchet.run/self-hosting/docker-compose](https://docs.hatchet.run/self-hosting/docker-compose) with Typescript results in the following error message:

```sh
> client@1.0.0 worker
> npx ts-node worker.ts

/Users/mbecker/apps/hatchet/client/node_modules/@hatchet-dev/typescript-sdk/clients/hatchet-client/hatchet-client.js:92
                throw new Error(`Invalid client config: ${e.message}`);
                      ^
Error: Invalid client config: [
  {
    "received": "none",
    "code": "invalid_enum_value",
    "options": [
      "tls",
      "mtls"
    ],
    "path": [
      "tls_config",
      "tls_strategy"
    ],
    "message": "Invalid enum value. Expected 'tls' | 'mtls', received 'none'"
  }
]
```

The docker compose instances are running and available via browser. The env variables are in the `.env` file.

From the SKD`s error message I would say that a connection to a localhost withput any TLS doesn't work.

Seems to related to this issue.

P.s. The docs should be updated that the environment variables should be set directly or via `.env`-file.

My environment is as follows:

```json
{
  "name": "client",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "worker": "npx ts-node worker.ts"
  },
  "author": "",
  "license": "ISC",
  "dependencies": {
    "@hatchet-dev/typescript-sdk": "^0.1.27",
    "dotenv": "^16.4.5"
  }
}
```

```sh
(base) mbecker@mbecker-MBP client % node -v
v20.9.0
```

**Comment by grutt** [comment]
@steebchen good catch.

I'll look into this shortly. It looks like our docs are out of sync.

Our quick start repos use Caddy to manage TLS certs but you do need to specify a TLS ROOT CA FILE, mine is at:
`HATCHET_CLIENT_TLS_ROOT_CA_FILE="/Users/gabrielruttner/Library/Application Support/Caddy/pki/authorities/local/root.crt"`

https://github.com/hatchet-dev/hatchet-typescript-quickstart

@mbecker can you let me know if this unblocks you for now?

**Comment by mbecker** [comment]
Hi @grutt ,

Thank you for the feedback.

It seems that you have manually installed Caddy on your system, is that correct? Or, how do you have a Caddy Cert in your OS path?

For which image / container do you use the env variable `HATCHET_CLIENT_TLS_ROOT_CA_FILE`?

**Comment by mbecker** [comment]
Hi,

seems that any SDK (Typescript / Go) does not support the env variable `HATCHET_CLIENT_TLS_STRATEGY=none`. The setup is as documented at the following page [https://docs.hatchet.run/self-hosting/docker-compose](https://docs.hatchet.run/self-hosting/docker-compose):

- Every hatchet instances skip the cert creation (That's my guess from the docer compose yaml)
- The Caddy server proxies the hatchet frontend and api at http://localhot:XXX (NO HTTPS)

All client SDK can't connect to a hatchet instance with ONLY HTTP.

I've managed to run a Caddy server in the docker compose with a self signed certificated which is trusted by my local system. See [https://github.com/josue/docker-caddy-reverse-proxy](https://github.com/josue/docker-caddy-reverse-proxy) for more information.

The Typescript SDK can connect with `HATCHET_CLIENT_TLS_STRATEGY=tls`. But in the UI the worker isn't listed. Events which are sent successfully are note listed in the UI. The worker does not start processing any workflows / tasks.

The Go SDK can't connect with `HATCHET_CLIENT_TLS_STRATEGY=tls`. It does need (Must Have!) the following env variables:

```sh
HATCHET_CLIENT_TLS_ROOT_CA_FILE=xxx
HATCHET_CLIENT_TLS_CERT_FILE=xx
HATCHET_CLIENT_TLS_KEY_FILE=xxx
```

So, anyways, the Docker Compose examples doesn't seem to work.

I would suggest that all SDKs should support `HATCHET_CLIENT_TLS_STRATEGY=none`. Additionally, I would say the Go SDK should work with `HATCHET_CLIENT_TLS_STRATEGY=tls` and no other env variables.

**Comment by abelanger5** [comment]
There are a few issues here that I believe are unrelated. Fixing the original issue in #269, and thanks for the report! 

To respond to some of the other issues:

> seems that any SDK (Typescript / Go) does not support the env variable HATCHET_CLIENT_TLS_STRATEGY=none

~Only the Go SDK is missing this strategy~ Never mind, seeing the enum issue in TS as well. Will fix in a separate PR. ~Both Typescript and Python have support for this and have been tested, but happy to debug more specifically if you're seeing something different.~

> The Typescript SDK can connect with HATCHET_CLIENT_TLS_STRATEGY=tls. But in the UI the worker isn't listed. Events which are sent successfully are note listed in the UI. The worker does not start processing any workflows / tasks.

Just to double check, are you sure you're viewing the correct tenant in the UI? If events are being sent without errors (regardless of the worker/workflow run status) they will show up in the UI. If you've double checked this, could you create a new issue and share any relevant logs from the engine or the API? 

> Additionally, I would say the Go SDK should work with HATCHET_CLIENT_TLS_STRATEGY=tls and no other env variables.

It does work without these things if your cert chain is trusted by your machine. I have a setup with a Caddy-signed cert that only requires the API token to be set, as the `tls` strategy is the default. Happy to document how I got this running in detail, you shouldn't need `HATCHET_CLIENT_TLS_CERT_FILE` or `HATCHET_CLIENT_TLS_KEY_FILE` in the default case. If you send these certificates, the server will check them anyway, but they are not required in the `tls` strategy case.

**Comment by steebchen** [comment]
I think the point here is that we need to get rid of any manual TLS related things such as setting up TLS related env vars, especially as the quickstart does not document this. In general, requiring Caddy to run a simple example is not a great UX.

**Comment by mbecker** [comment]
Hi,

Just wanted to paste the following TS SDK line as I've seen your edit of the comment :-)

[https://github.com/hatchet-dev/hatchet-typescript/blob/20094b1c95695f10f0cd3e941d8b6c4c97f8dd57/src/clients/hatchet-client/client-config.ts#L5](https://github.com/hatchet-dev/hatchet-typescript/blob/20094b1c95695f10f0cd3e941d8b6c4c97f8dd57/src/clients/hatchet-client/client-config.ts#L5)

But yes, it's a different issue.

I'm sure that the Docker Compose with a Caddy plus a local cert from me which is trusted by the OS is working. I would be happy to see your setup with a local Caddy rather a Docker Caddy. Thanks!

**Comment by abelanger5** [comment]
I should have read the comments above more carefully, but to clarify - the quickstart examples [here](https://docs.hatchet.run/self-hosting/docker-compose) do not use Caddy TLS, they just use HTTP as @mbecker mentioned. Additionally, the quickstart examples are **only reverse proxying Caddy to the frontend and API, not the engine**. So anything that gets done with Caddy modified from this example is not changing TLS for the engine unless add an additional block to the Caddyfile, for example:

```
app.dev.hatchet-tools.com {
tls internal

	handle /api/* {
		reverse_proxy localhost:8080
	}

	handle /* {
		reverse_proxy localhost:5173
	}
}

grpc.dev.hatchet-tools.com {
    tls internal

    reverse_proxy h2c://127.0.0.1:7070
}
```  

Once you add a block like this, now you need a CA cert trusted by the system to connect to the engine, or you need to explicitly set it via an environment variable. I mentioned that I have Caddy TLS running locally which is also how @grutt has configured his local environment, and you can do this by setting something like `HATCHET_CLIENT_TLS_ROOT_CA_FILE="/Users/gabrielruttner/Library/Application Support/Caddy/pki/authorities/local/root.crt"`. You'll also need to change the `SERVER_GRPC_BROADCAST_ADDRESS` to a value like `grpc.dev.hatchet-tools.com:443` (in this case). 

But to reiterate, **users should not need to set any cert files and that's not expected behavior for the quickstarts**. So there are two issues here - one that the Go SDK doesn't support this option (fixed in the linked PR), the other that the Typescript SDK throws an error when this env var is set. 

> I think the point here is that we need to get rid of any manual TLS related things such as setting up TLS related env vars, especially as the quickstart does not document this. In general, requiring Caddy to run a simple example is not a great UX.

One note on this, I think requiring that users set the environment variable `HATCHET_CLIENT_TLS_STRATEGY=none` is reasonable to make TLS explicitly opt-out. 

And I agree on the Caddy front - we should probably just bundle a "lite" docker image which serves the API, frontend and engine for local setups and quickstart. This image could even use Caddy internally.
